modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/core/components/PageRenderer/Layout.es.js

```diff
@@ -29,6 +31,35 @@ export function Layout({components, editable, itemPath, rows, viewMode}) {
 
 	const Components = components ?? mergeVariants(editable, variants);
 
+	useEffect(() => {
+		const undoHandler = () => {
+			dispatch({type: EVENT_TYPES.HISTORY.PREV});
+		};
+		const redoHandler = () => {
+			dispatch({type: EVENT_TYPES.HISTORY.NEXT});
+		};
+
+		Liferay.on('journal:undo', undoHandler);
+		Liferay.on('journal:redo', redoHandler);
+
+		return () => {
+			Liferay.detach('journal:undo', undoHandler);
+			Liferay.detach('journal:redo', redoHandler);
+		};
+	}, [dispatch]);
+
+	useEffect(() => {
+		const handleStoreState = () => {
+			dispatch({type: EVENT_TYPES.HISTORY.ADD});
+		};
+
+		Liferay.on('journal:storeState', handleStoreState);
+
+		return () => {
+			Liferay.detach('journal:storeState', handleStoreState);
+		};
+	}, [dispatch]);
```

Comment:
since these `useEffect` have the same depedency `dispatch` couldn't they be only one `useEffect` instead of two?